### PR TITLE
GString: Avoid copy for From<GString> for String impl

### LINKED ISSUE #422: GString: erroneous string copies in current implementation
```
[13:17:42] <slomo> philn: From<GString> for String currently copies the CString (in case of ForeignOwned), instead of just converting the CString into a String without copying
[13:18:57] <slomo> philn: similarly impl From<String> for GString can probably also happen without copying
```

### PR DESCRIPTION
**Pull request by philn**
Fixes #422

### CODE REVIEW

File: src/gstring.rs [hunk 1]
```diff
@@ -180,7 +180,16 @@ impl Deref for GString {
...
-        String::from(s.as_str())
+        if let GString::ForeignOwned(ref cstring) = s {
+            if let Ok(s) = String::from_utf8(cstring.as_bytes().to_vec()) {
+                s
+            } else {
+                String::from(s.as_str())
+            }
+        } else {
+            String::from(s.as_str())
+
```
Comment (EPashkin) [inline src/gstring.rs]: Unneeded empty line.

**Comment by EPashkin** [comment]
@philn Thansk, LGFM
Only possible change it do early return in nocopy case and use one call `String::from(s.as_str())`
but IMHO rust can optimize it yourself

**Comment by philn** [comment]
Thanks, I updated the PR, code looks simpler now indeed!

File: src/gstring.rs [hunk 2]
```diff
@@ -180,6 +180,11 @@ impl Deref for GString {
 impl From<GString> for String {
     #[inline]
     fn from(s: GString) -> Self {
+        if let GString::ForeignOwned(ref cstring) = s {
+            if let Ok(s) = String::from_utf8(cstring.as_bytes().to_vec()) {
```
Comment (EPashkin) [inline src/gstring.rs]: Hm, I missed that we have buildin method
https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_string

Comment (philn) [reply src/gstring.rs]: Sorry, I don't follow.

Comment (philn) [reply src/gstring.rs]:
```
error[E0507]: cannot move out of borrowed content                                                                                                                                                                  
   --> src/gstring.rs:184:28                                                                                                                                                                                       
    |                                                                                                                                                                                                              
184 |             if let Ok(s) = cstring.into_string() {                                                                                                                                                           
    |                            ^^^^^^^ cannot move out of borrowed content                                                                                                                                       
                                                                                                                                                                                                                   
error: aborting due to previous error 
```

Comment (EPashkin) [reply src/gstring.rs]: We want remove copy in foreign owned case.
As [`slice::to_vec`](https://doc.rust-lang.org/std/primitive.slice.html#method.to_vec) make copy,
we need get inner slice and give it to new String with transfer ownership.

Comment (sdroege) [reply src/gstring.rs]: Yeah, remove the `ref` keyword here

Comment (philn) [reply src/gstring.rs]: That's not possible because GString implements the Drop trait :/

Comment (sdroege) [reply src/gstring.rs]: Then let's do the standard workaround for that :) Store the `CString` in an `Option<_>` and use `Option::take()` here, and in all other places use `as_ref().unwrap()` when accessing the `Option<_>`

File: src/gstring.rs [hunk 3]
```diff
@@ -48,7 +48,7 @@ impl GString {
                 CStr::from_bytes_with_nul_unchecked(bytes)
             }
             GString::ForeignOwned(cstring) => {
-                cstring.as_c_str()
+                cstring.as_ref().unwrap().as_c_str()
```
Comment (EPashkin) [inline src/gstring.rs]: Maybe use `expect`?

Comment (GuillaumeGomez) [reply src/gstring.rs]: > Maybe use `expect`?

I approve!

**Comment by philn** [comment]
CI is green and comments addressed :)

**Comment by EPashkin** [comment]
@philn Thanks, 👍

**Comment by GuillaumeGomez** [comment]
Thanks!

**Pull request merged**